python/cog/server/clients.py

```diff
@@ -0,0 +1,319 @@
+            log.info("sending webhook with status %s", response["status"])
+            # For other requests, don't retry, and ignore any errors
+            try:
+                await self.webhook_client.post(url, json=response)
+            except httpx.RequestError:
+                log.warn("caught exception while sending webhook", exc_info=True)
+
+    def make_webhook_sender(
+        self, url: Optional[str], webhook_events_filter: Collection[WebhookEvent]
+    ) -> WebhookSenderType:
+        throttler = ResponseThrottler(response_interval=_response_interval)
+
+        async def sender(response: Any, event: WebhookEvent) -> None:
+            if url and event in webhook_events_filter:
+                if throttler.should_send_response(response):
+                    await self.send_webhook(url, response, event)
+                    throttler.update_last_sent_response_time()
+
+        return sender
+
+    # files
+
+    async def upload_file(
+        self, fh: io.IOBase, *, url: Optional[str], prediction_id: Optional[str]
+    ) -> str:
+        """put file to signed endpoint"""
+        log.debug("upload_file")
+
+        fh.seek(0)
+
+        # try to guess the filename of the given object
+        name = getattr(fh, "name", "file")
+        filename = os.path.basename(name) or "file"
+        assert isinstance(filename, str)
+
+        guess, _ = mimetypes.guess_type(filename)
+        content_type = guess or "application/octet-stream"
+
+        # this code path happens when running outside replicate without upload-url
+        # in that case we need to return data uris
+        if url is None:
+            return file_to_data_uri(fh, content_type)
+        assert url
+
+        # ensure trailing slash
+        url_with_trailing_slash = url if url.endswith("/") else url + "/"
+
+        url = url_with_trailing_slash + filename
+
+        headers = {"Content-Type": content_type}
+        if prediction_id is not None:
+            headers["X-Prediction-ID"] = prediction_id
+
+        # this is a somewhat unfortunate hack, but it works
+        # and is critical for upload training/quantization outputs
+        # if we get multipart uploads working or a separate API route
+        # then we could drop this
+        if url and (".internal" in url or ".local" in url):
+            log.info("doing test upload to %s", url)
+            resp1 = await self.file_client.put(
+                url,
+                content=b"",
+                headers=headers,
+                follow_redirects=False,
+            )
+            if resp1.status_code == 307 and resp1.headers["Location"]:
+                log.info("got file upload redirect from api")
+                url = resp1.headers["Location"]
+
+        log.info("doing real upload to %s", url)
+        # set connect timeout to slightly more than a multiple of 3 to avoid
+        # aligning perfectly with TCP retransmission timer
+        timeout = httpx.Timeout(10.0, read=15.0)
```

Comment:
Isn't this already configured on the `file_client`?

Reply:
yeah, I'm not sure what @mattt's motivation was adding this in https://github.com/replicate/cog/commit/3ccc7644b52fa5331584a7be0cf2f9cf40419216#diff-a2784bee5e78b50b83f93560b26adb124c350aad2ddfe5a98699ec5f114c4f06R240-R242

Reply:
Sorry, that was a bad rebase.